Refetch should not happen if no active subscribers

AlexanderArvidsson opened the pull request:
Fixes #1530

All logic that checks if there are active subscriptions in the code does a falsey check as well as a keys length check on the subscription sub state, except for the code that refetches valid queries. 
[Here is an example from the handleUnsubscribe function.](https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/core/buildMiddleware/cacheCollection.ts#L98)
[Here is another from invalidateTags.](https://github.com/reduxjs/redux-toolkit/blob/d4e95caa26ea1c76a1ecead8f09b1897b4444d40/packages/toolkit/src/query/core/buildMiddleware/invalidationByTags.ts#L75)
This PR includes the same check to resolve this issue. See #1530 for more information.
Attempting a refetch for data that has no subscribers should remove it from the cache, similar to `invalidateTags`, because the data would become stale relative to other cache data which would be refetched on focus / reconnect.

phryneas commented:
Ha, timing ^^

What I said in the other thread pretty much applies here - "not fetching" in that case is not really an option, as that would leave stale data in the cache. In this case, the data would need to be removed from the cache, like it is in the case of cache invalidation. 
Could you please adjust the PR accordingly?

AlexanderArvidsson commented:
I will adjust the PR tomorrow with the fixes, as per our discussion in the issue.

AlexanderArvidsson commented:
@phryneas I changed the logic now and wrote a test for it. I considered placing the logic in `cleanup.test.tsx` but since `refetchValidQueries` is only run when refetching on focus or reconnect, I added it there.

Is it necessary to also include the same test for reconnect? They run the same function, so if the normal tests for reconnect passes, and my new test passes, then it should be guaranteed to pass on reconnect too.

AlexanderArvidsson commented on packages/toolkit/src/query/core/buildMiddleware/windowEventHandling.ts:
```diff
@@ -35,24 +39,27 @@ export const build: SubMiddlewareBuilder = ({
+            })
+          )
+        } else if (querySubState.status !== QueryStatus.uninitialized) {
+          const shouldRefetch =
+            Object.values(subscriptionSubState).some(
+              (sub) => sub[type] === true
+            ) ||
+            (Object.values(subscriptionSubState).every(
+              (sub) => sub[type] === undefined
+            ) &&
+              state.config[type])
+
+          if (shouldRefetch) {
+            api.dispatch(refetchQuery(querySubState, queryCacheKey))
+          }
```
Should `removeQueryResult` be guarded by `shouldRefetch` as well? What do you think @phryneas ?

phryneas replied:
Yes, moving it into that guard might be a good way. That way we remove as little as necessary from the cache.

phryneas commented on packages/toolkit/src/query/core/buildMiddleware/windowEventHandling.ts:
```diff
@@ -35,24 +39,27 @@ export const build: SubMiddlewareBuilder = ({
-
-        if (shouldRefetch) {
-          api.dispatch(refetchQuery(querySubState, queryCacheKey))
+        if (!subscriptionSubState || !querySubState) return
```
On re-reading this... shouldn't this `return` always have been a `continue`?

AlexanderArvidsson replied:
Very good catch, I'm surprised this hasn't caused more severe issues. I'll adjust it soon.

phryneas commented:
I'l try tol take a look at it this evening or tomorrow.
Afaik, @markerikson wanted to do a patch release soon so we should be able to get this in.

phryneas commented:
Looks good to me. Let's get this in.

phryneas commented:
Released in https://github.com/reduxjs/redux-toolkit/releases/tag/v1.7.2
